### PR TITLE
First step to supporting access to sysrepo across multiple containers

### PR DESCRIPTION
**Pull request by imiller0**
Convert all shm_open and shm_unlink calls to open() and unlink().  This allows the sysrepo shared-memory directory to be a shared-memory filesystem mounted from "host" to container.  This is the first step towards allowing full cooperative access to sysrepo by multiple containers. 

By using open() and unlink(), all handles to the shared memory objects (main/ext files, subscriptions, etc) can all use absolute path references.

A quick look at the glibc implementations of shm_open and shm_unlink show a couple notable differences, all of which I believe don't impact sysrepo:
- shm_open adds the O_NOFOLLOw flag to the open() call.  This flag causes a failure if the opened "file" is a symbolic link.  None of the sysrepo files in shared memory are (as far as I can tell) symbolic links.
- shm_open adds the O_CLOEXEC flag to the open() call.  This flag closes the file descriptor across execve system calls (forked child process).  Running "nm" on the libsysrepo and libyang libraries does not reveal any exec* calls.
- shm_open searches the filesystem for a shared memory mount point.  The search starts at /dev/shm then moves on to mounts in /proc/mounts or /etc/fstab. For sysrepo the mount point is specified in common.h.in, so the search in this case is not beneficial.
- shm_open verifies the filesystem is a shared memory filesystem by checking the superblock "magic" number (for SHMFS or TMPFS).  This validation will be lost by this change-set, but the #define in common.h.in is set to /dev/shm which should always be a shared memory filesystem.
- shm_unlink() performs the same filesystem search as shm_open to locate the path, but is otherwise just a passthrough to unlink().

Please let me know if you have any questions or if anything needs to be re-worked.  Other changes for allowing access across container boundaries will be pushed "soon".
Thanks!

### CODE REVIEW

**Comment by michalvasko** [comment]
Hi,
thanks for splitting the changes into separate commits and PRs, it will be easier to review it this way. Also, it seems you dug deep into `glibc`, which is great.

> shm_open adds the O_NOFOLLOw flag to the open() call. This flag causes a failure if the opened "file" is a symbolic link. None of the sysrepo files in shared memory are (as far as I can tell) symbolic links.

They are not and they should not be. So actually always using this flag seems like a reasonable security measure to me. And it can probably be used for all the calls, even when opening `startup` or YANG files as all of those should be managed exclusively by *sysrepo*.

> shm_open adds the O_CLOEXEC flag to the open() call. This flag closes the file descriptor across execve system calls (forked child process). Running "nm" on the libsysrepo and libyang libraries does not reveal any exec* calls.

Did not think about this flag but now that you mention it, it should also be used for all the calls. Or at least for opening the main SHM when a new connection is created because forking then causes a fatal problem (and it is mentioned in the docs). This way *sysrepo* would simply stop working, which is the most that can be done.

So, ideally there could be a new macro defined with these 2 flags and used in all the `open()` calls.

> shm_open searches the filesystem for a shared memory mount point. The search starts at /dev/shm then moves on to mounts in /proc/mounts or /etc/fstab. For sysrepo the mount point is specified in common.h.in, so the search in this case is not beneficial.

I did not know this and a static path is used because the SHM files are inspected also using `access()` and I did not know how else to implement it, about any portable way of learning the SHM directory. But there were no issues with this it seems and I am not sure how would you go about the functionality in this PR if it worked any other way, anyway. But if you have some ideas, let me know.

> shm_open verifies the filesystem is a shared memory filesystem by checking the superblock "magic" number (for SHMFS or TMPFS). This validation will be lost by this change-set, but the #define in common.h.in is set to /dev/shm which should always be a shared memory filesystem.

Well, shared memory is actually used mostly for performance reasons and the feature of it not being persistent is just an additional useful perk. So, this should not be a problem, I think. With connection to the previous bullet, it may be agood idea to make the SHM directory a `cmake` variable for some specific cases of SHM being mounted in some other path or even perhaps not being supported on some platforms.

> shm_unlink() performs the same filesystem search as shm_open to locate the path, but is otherwise just a passthrough to unlink().

Right, if everything needed was performed when creating the file, it should be fine to remove it without all the checks.

**Comment by imiller0** [comment]
> So, ideally there could be a new macro defined with these 2 flags and used in all the `open()` calls.

Done.  I've added a macro SR_SHM_OPEN which is now used where the previous shm_open() calls were made.  This can be extended to other open() calls if/as needed.

> 
> > shm_open searches the filesystem for a shared memory mount point. The search starts at /dev/shm then moves on to mounts in /proc/mounts or /etc/fstab. For sysrepo the mount point is specified in common.h.in, so the search in this case is not beneficial.
> 
> I did not know this and a static path is used because the SHM files are inspected also using `access()` and I did not know how else to implement it, about any portable way of learning the SHM directory. But there were no issues with this it seems and I am not sure how would you go about the functionality in this PR if it worked any other way, anyway. But if you have some ideas, let me know.
> 
> > shm_open verifies the filesystem is a shared memory filesystem by checking the superblock "magic" number (for SHMFS or TMPFS). This validation will be lost by this change-set, but the #define in common.h.in is set to /dev/shm which should always be a shared memory filesystem.
> 
> Well, shared memory is actually used mostly for performance reasons and the feature of it not being persistent is just an additional useful perk. So, this should not be a problem, I think. With connection to the previous bullet, it may be agood idea to make the SHM directory a `cmake` variable for some specific cases of SHM being mounted in some other path or even perhaps not being supported on some platforms.

In  a subsequent change set I have a change which makes the shm path configurable via an environment variable.  This allows different paths to be used by different containers and/or the host.

**Comment by michalvasko** [comment]
> Done. I've added a macro SR_SHM_OPEN which is now used where the previous shm_open() calls were made. This can be extended to other open() calls if/as needed.

Right, I was too vague, I guess. My idea was to simply define something like `#define SR_OPEN_FLAGS O_NOFOLLOW | O_CLOEXEC` and use it in every `open()` call, but what you wrote is shorter. So yes, please finish the change completely and use it for all the `open()` calls, no reason not to.

> In a subsequent change set I have a change which makes the shm path configurable via an environment variable. This allows different paths to be used by different containers and/or the host.

Alright, that is what I suggested.

File: src/executables/sysrepo-plugind.c [hunk 1]
```diff
@@ -196,7 +196,7 @@ daemon_init(int debug, sr_log_level_t log_level)
     }
 
     /* redirect standard files to /dev/null */
-    fd = open("/dev/null", O_RDWR, 0);
+    fd = open("/dev/null", O_RDWR | O_NOFOLLOW, 0);
```
Comment (michalvasko) [inline src/executables/sysrepo-plugind.c:199]: I really hope `/dev/null` is not some malicious symlink :) However, are you sure it cannot be a symbolic link on some regular distribution?

File: src/common.h.in [hunk 2]
```diff
@@ -93,6 +93,9 @@ int pthread_mutex_timedlock(pthread_mutex_t *mutex, const struct timespec *absti
 /** macro for checking session type */
 #define SR_IS_EVENT_SESS(session) (session->ev != SR_SUB_EV_NONE)
 
+/** macro for opening files in SHM */
```
Comment (michalvasko) [inline src/common.h.in:96]: Update the description please. Best to mention that all files are created by sysrepo, so there can be no symlinks and forking once a sysrepo connection is created is forbidden.

**Review by michalvasko — CHANGES_REQUESTED** [review]
Looks almost the way I wanted, just some final minor problems.

File: src/executables/sysrepo-plugind.c (hunk 1 above)
Comment (imiller0) [reply src/executables/sysrepo-plugind.c:199]: Good point.  No need for the restriction here.

File: src/common.h.in (hunk 2 above)
Comment (imiller0) [reply src/common.h.in]: Done. :)

**Review by michalvasko — APPROVED** [review]

**Pull request merged**